Sandman hides apoptosis null-node failure as success

## Context

Operator ran `npm run ai:run-sandman` on 2026-05-20. The run reached GraphMaintenanceService vector apoptosis, detected 8344 orphaned nodes, then logged a DreamService error while still printing `✅ Sandman cycle complete.` and exiting with code 0.

Observed stack:

```text
[INFO] [GraphMaintenanceService] Apoptosis detected 8344 orphaned nodes. Commencing eradication...
[ERROR] [DreamService] Failed to process undigested sessions: TypeError: Cannot read properties of null (reading 'id')
    at Store.getKey (src/data/Store.mjs:554:25)
    at Store.splice (src/collection/Base.mjs:1483:45)
    at Store.splice (ai/graph/Store.mjs:161:30)
    at Store.remove (src/collection/Base.mjs:1396:14)
    at Database.removeNode (ai/graph/Database.mjs:413:18)
    at GraphService.mjs:986:43
    at Array.forEach (<anonymous>)
    at GraphService.mjs:986:21
    at Database.transaction (ai/graph/Database.mjs:474:13)
    at GraphService.removeNodes (ai/services/memory-core/GraphService.mjs:985:17)
...
✅ Sandman cycle complete.
Process finished with exit code 0
```

## The Problem

The REM maintenance path can fail during apoptosis deletion and still look successful to the caller. This is dangerous for operator runs and future automation because the process-level success signal no longer means the REM cycle completed.

There are two coupled failure surfaces:

1. `GraphService.removeNodes()` can pass an invalid/null node id into `Database.removeNode()` during apoptosis cleanup.
2. `DreamService.processUndigestedSessions()` catches the error, logs it, does not rethrow or return a failure result, and `buildScripts/ai/runSandman.mjs` then prints success and sets `process.exitCode = 0`.

## The Architectural Reality

- `ai/daemons/services/GraphMaintenanceService.mjs:49-53` calls `GraphService.getOrphanedNodes()` and then `GraphService.removeNodes(orphaned)`.
- `ai/services/memory-core/GraphService.mjs:952-975` selects orphan rows from SQLite `Nodes` and pushes `row.id` into the deletion list without validating the id.
- `ai/services/memory-core/GraphService.mjs:982-988` wraps `nodeIds.forEach(id => this.db.removeNode(id))` in `Database.transaction()`.
- `ai/graph/Database.mjs:407-414` calls `me.nodes.remove(nodeId)` directly.
- `src/collection/Base.mjs:1294-1297` treats `null` as an item because `typeof null === 'object'`.
- `src/data/Store.mjs:542-554` then calls `item[keyProperty]`, which throws for `item === null`.
- `ai/daemons/DreamService.mjs:248-254` catches and logs the failure without propagating it.
- `buildScripts/ai/runSandman.mjs:218-224` awaits `DreamService.processUndigestedSessions()`, then unconditionally synthesizes Golden Path, prints success, and sets exit code 0 if no exception escapes.
- `ai/graph/storage/SQLite.mjs:83-87` declares `Nodes.id TEXT PRIMARY KEY` without explicit `NOT NULL`; existing SQLite rowid-table semantics can allow historical null primary-key rows unless explicitly constrained or guarded at insert time.

## The Fix

Implement a narrow hardening pass across the owning boundaries:

1. Add validation/repair handling so GraphService apoptosis never calls `Database.removeNode(null)` or otherwise routes invalid ids through `Store.remove()`.
2. Prevent future invalid graph node ids from entering SQLite through `SQLite.addNodes()` / graph upsert paths, with an explicit diagnostic rather than a latent corrupt row.
3. Ensure `DreamService.processUndigestedSessions()` exposes failure to callers, either by rethrowing after logging or by returning a structured failure result consumed by `runSandman`.
4. Make `runSandman` fail non-zero whenever REM processing failed, while preserving the heavy-maintenance lease semantics and the held-lease early-exit behavior.
5. Add focused unit coverage for the null-id apoptosis guard and the Sandman/DreamService failure propagation contract.

## Contract Ledger Matrix

| Target Surface | Source of Authority | Proposed Behavior | Fallback | Docs | Evidence |
|---|---|---|---|---|---|
| `GraphService.removeNodes(nodeIds)` | This ticket + 2026-05-20 Sandman stack | Invalid/null ids are rejected or repaired before `Database.removeNode()`; no `Store.getKey(null)` TypeError | Emit explicit diagnostic naming invalid node ids / corrupt rows | JSDoc on `removeNodes()` and/or `getOrphanedNodes()` | Unit test passes `[null]` or corrupt orphan row and proves no `Store.getKey(null)` crash |
| SQLite node insert path | This ticket + schema audit | New node writes require a non-empty string id | Throw explicit invalid-node-id error before insert | JSDoc on `SQLite.addNodes()` | Unit test proves null/undefined node id is rejected before SQLite persistence |
| `DreamService.processUndigestedSessions()` | This ticket + runSandman caller contract | REM failures are observable to callers | Structured failure result if rethrow would break existing caller | JSDoc on method return/error contract | Unit test proves GraphMaintenanceService failure reaches caller |
| `npm run ai:run-sandman` | Operator CLI contract | Fatal REM failure exits non-zero and does not print successful completion | Held lease still exits 0 without mutation, as today | Script comments near exitCode handling | Script-level test or unitized dependency-injected test proves failure => exitCode 1 |

## Acceptance Criteria

- [ ] Regression test reproduces the `Store.getKey(null)` path from apoptosis deletion and verifies the new guard prevents the TypeError.
- [ ] Invalid/null graph node ids cannot be inserted through the normal SQLite `addNodes()` persistence path without an explicit error.
- [ ] `DreamService.processUndigestedSessions()` no longer silently swallows a GraphMaintenanceService failure from callers.
- [ ] `buildScripts/ai/runSandman.mjs` exits non-zero and does not print `✅ Sandman cycle complete.` when REM processing fails.
- [ ] Held heavy-maintenance lease behavior remains unchanged: held lease exits cleanly without running REM or decay.
- [ ] Post-merge validation: rerun `npm run ai:run-sandman`; the previous `Cannot read properties of null (reading 'id')` error is gone, and any future fatal REM error returns non-zero.

## Out of Scope

- Manually deleting or rewriting all current orphan rows outside the tested repair/guard path.
- Reworking vector-dimension mismatch configuration (`gemini-embedding-001` 3072 vs configured 4096); that warning is visible in the same log but is a separate configuration/collection-dimension concern.
- Changing apoptosis retention policy or protected node labels.
- Reworking heavy-maintenance lease acquisition semantics.

## Avoided Traps

- **Treating the exit code 0 as success** — rejected. The logged DreamService error falsifies the success claim.
- **Folding this into #11595** — rejected. #11595 fixed string-shaped rollback payloads; this stack reaches `Store.getKey(null)` before that rollback failure class.
- **Only changing `Collection.Base.isItem(null)`** — rejected as an insufficient sole fix. It may avoid this TypeError, but it does not explain or prevent invalid graph ids entering apoptosis or fix the swallowed REM failure.
- **Only skipping null ids in apoptosis** — rejected as incomplete. It leaves Sandman able to hide future fatal REM failures behind exit code 0.

## Related

- #11595 / PR #11611 — adjacent prior apoptosis rollback shape bug, closed/merged.
- `ai/daemons/services/GraphMaintenanceService.mjs:49-53`
- `ai/services/memory-core/GraphService.mjs:952-988`
- `ai/graph/Database.mjs:407-414`
- `src/collection/Base.mjs:1294-1297`
- `src/data/Store.mjs:542-554`
- `ai/daemons/DreamService.mjs:248-254`
- `buildScripts/ai/runSandman.mjs:218-224`

Origin Session ID: d13c94dd-e721-4e28-ac9e-4d0b3c0f66de
Retrieval Hint: `query_raw_memories("Sandman apoptosis null node Store.getKey exit code 0 DreamService")`
Retrieval Hint: `query_raw_memories("GraphService removeNodes null id runSandman success after error")`

--- a/ai/daemons/DreamService.mjs
+++ b/ai/daemons/DreamService.mjs
@@ -251,6 +251,7 @@ class DreamService extends Base {
             logger.info('[DreamService] REM pipeline completed.');
         } catch (error) {
             logger.error('[DreamService] Failed to process undigested sessions:', error);
+            throw error;
         } finally {
             this.isProcessing = false;
         }

--- a/ai/graph/Database.mjs
+++ b/ai/graph/Database.mjs
@@ -405,6 +405,10 @@ class Database extends Base {
      * @param {String} nodeId
      */
     removeNode(nodeId) {
+        if (typeof nodeId !== 'string' || nodeId.length === 0) {
+            throw new Error(`Graph Database removeNode requires a non-empty string node id. Received: ${String(nodeId)}`);
+        }
+
         let me            = this,
             outbound      = me.edges.getByIndex('source', nodeId),
             inbound       = me.edges.getByIndex('target', nodeId),

--- a/ai/graph/storage/SQLite.mjs
+++ b/ai/graph/storage/SQLite.mjs
@@ -167,6 +167,11 @@ class SQLite extends Base {
                     label: isRecord ? node.get('label') : node.label,
                     properties: isRecord ? node.get('properties') : node.properties
                 };
+
+                if (typeof nodeData.id !== 'string' || nodeData.id.length === 0) {
+                    throw new Error(`SQLite graph node writes require a non-empty string id. Received: ${String(nodeData.id)}`);
+                }
+
                 stmt.run(nodeData.id, nodeData.properties?.userId || null, JSON.stringify(nodeData));
             }
         });

--- a/ai/services/memory-core/GraphService.mjs
+++ b/ai/services/memory-core/GraphService.mjs
@@ -34,6 +34,16 @@ function isRlsVisible(entity, requesterUserId) {
 }
 
 /**
+ * Validates the Native Edge Graph node-id contract before writes/deletes reach
+ * the lower Store layer, where `null` otherwise looks object-like.
+ * @param {*} id Candidate graph node id.
+ * @returns {Boolean} true for non-empty string ids.
+ */
+function isValidGraphNodeId(id) {
+    return typeof id === 'string' && id.length > 0;
+}
+
+/**
  * @summary Service that manages the SQLite Knowledge Graph (Nodes and Edges).
  *
  * It provides the topological layout of the Neo.mjs namespace, knowledge,
@@ -177,6 +187,10 @@ class GraphService extends Base {
      * @param {Object} nodeData
      */
     upsertNode({id, type, name, description, semanticVectorId, state, updatedAt, properties}) {
+        if (!isValidGraphNodeId(id)) {
+            throw new Error(`[GraphService] Cannot upsert graph node without a non-empty string id. Received: ${String(id)}`);
+        }
+
         // Lazy-load from SQLite before in-memory check — prevents cold-cache stubs from
         // overwriting rich SQLite rows via addNodes' ON CONFLICT DO UPDATE semantics.
         // Mirrors the discipline in getNode:424. Resolves #10230.
@@ -981,6 +995,12 @@ class GraphService extends Base {
      */
     removeNodes(nodeIds) {
         if (!nodeIds || nodeIds.length === 0) return;
+
+        const invalidNodeIds = nodeIds.filter(id => !isValidGraphNodeId(id));
+
+        if (invalidNodeIds.length > 0) {
+            throw new Error(`[GraphService] removeNodes received invalid node id(s): ${invalidNodeIds.map(id => String(id)).join(', ')}`);
+        }
 
         this.db.transaction(() => {
             nodeIds.forEach(id => this.db.removeNode(id));

--- a/buildScripts/ai/runSandman.mjs
+++ b/buildScripts/ai/runSandman.mjs
@@ -215,11 +215,7 @@ export async function runSandman() {
 
                 console.log('✅ Services Ready. Entering REM Sleep...');
 
-                // Execute the REM pipeline (extract undigested graph entities + Golden Path synthesis)
-                await DreamService.processUndigestedSessions();
-                await GoldenPathSynthesizer.synthesizeGoldenPath();
-
-                console.log('✅ Sandman cycle complete.');
+                await runRemPipeline();
                 process.exitCode = 0;
                 return {providerReady: true};
             } catch (e) {
@@ -264,6 +260,27 @@ export async function runSandman() {
     // above for the release-timing invariant). Nothing further to do here besides honoring
     // the inner exitCode contract.
     process.exit(process.exitCode);
+}
+
+/**
+ * Runs the Sandman REM cycle body after provider and service readiness gates pass.
+ * Exposed for unit coverage so fatal DreamService failures cannot regress into
+ * a misleading success log or successful process exit.
+ * @param {Object} options
+ * @param {Object} [options.dreamService=DreamService]
+ * @param {Object} [options.goldenPathSynthesizer=GoldenPathSynthesizer]
+ * @param {Object} [options.output=console]
+ * @returns {Promise<void>}
+ */
+export async function runRemPipeline({
+    dreamService          = DreamService,
+    goldenPathSynthesizer = GoldenPathSynthesizer,
+    output                = console
+} = {}) {
+    await dreamService.processUndigestedSessions();
+    await goldenPathSynthesizer.synthesizeGoldenPath();
+
+    output.log('✅ Sandman cycle complete.');
 }
 
 if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {

--- a/test/playwright/unit/ai/buildScripts/runSandman.spec.mjs
+++ b/test/playwright/unit/ai/buildScripts/runSandman.spec.mjs
@@ -31,6 +31,8 @@ import * as core      from '../../../../../src/core/_export.mjs';
  * provider-unavailable state from missing DreamService / Golden Path output.
  */
 test.describe('runSandman.mjs provider readiness diagnostics (#10587)', () => {
+    test.describe.configure({mode: 'serial'});
+
     let aiConfig;
     let logger;
     let runSandmanModule;
@@ -119,5 +121,27 @@ test.describe('runSandman.mjs provider readiness diagnostics (#10587)', () => {
         expect(logLines).toContain('mlx-community/test-model');
         expect(logLines).toContain('Start the configured OpenAI-compatible / MLX provider');
         expect(logLines).not.toContain('must-not-be-logged');
+    });
+
+    test('runRemPipeline propagates REM failures without printing success (#11698)', async () => {
+        const logs = [];
+
+        await expect(runSandmanModule.runRemPipeline({
+            dreamService: {
+                processUndigestedSessions: async () => {
+                    throw new Error('simulated REM failure');
+                }
+            },
+            goldenPathSynthesizer: {
+                synthesizeGoldenPath: async () => {
+                    throw new Error('Golden Path must not run after REM failure');
+                }
+            },
+            output: {
+                log: message => logs.push(message)
+            }
+        })).rejects.toThrow('simulated REM failure');
+
+        expect(logs.some(message => message.includes('Sandman cycle complete'))).toBe(false);
     });
 });

--- a/test/playwright/unit/ai/daemons/DreamService.spec.mjs
+++ b/test/playwright/unit/ai/daemons/DreamService.spec.mjs
@@ -778,6 +778,42 @@ test.describe('Neo.ai.services.memory-core.DreamService', () => {
         }
     });
 
+    test('processUndigestedSessions rethrows garbage-collection failures (#11698)', async () => {
+        const aiConfig = (await import('../../../../../ai/mcp/server/memory-core/config.mjs')).default;
+
+        const orig = {
+            provider      : aiConfig.modelProvider,
+            findUndigested: DreamService.findUndigestedSessions,
+            runGarbageCol : DreamService.runGarbageCollection,
+            loggerError   : logger.error,
+            isProcessing  : DreamService.isProcessing
+        };
+
+        const errors = [];
+
+        try {
+            aiConfig.modelProvider    = 'mock-provider';
+            DreamService.isProcessing = false;
+
+            DreamService.findUndigestedSessions = async () => [];
+            DreamService.runGarbageCollection   = async () => {
+                throw new Error('simulated apoptosis failure');
+            };
+            logger.error = (...args) => { errors.push(args); };
+
+            await expect(DreamService.processUndigestedSessions()).rejects.toThrow('simulated apoptosis failure');
+
+            expect(DreamService.isProcessing).toBe(false);
+            expect(errors.some(args => args[0] === '[DreamService] Failed to process undigested sessions:')).toBe(true);
+        } finally {
+            aiConfig.modelProvider              = orig.provider;
+            DreamService.findUndigestedSessions = orig.findUndigested;
+            DreamService.runGarbageCollection   = orig.runGarbageCol;
+            logger.error                        = orig.loggerError;
+            DreamService.isProcessing           = orig.isProcessing;
+        }
+    });
+
     test('synthesizeGoldenPath should mathematically select and inject Golden Path while rejecting BLOCKS', async () => {
         // Mock StorageRouter to return deterministic ChromaDB metric formats
         const originalGetSummary = StorageRouter.getSummaryCollection;

--- a/test/playwright/unit/ai/graph/Database.spec.mjs
+++ b/test/playwright/unit/ai/graph/Database.spec.mjs
@@ -97,6 +97,12 @@ test.describe('Neo.ai.graph.Database', () => {
         expect(db.edges.getCount()).toBe(0); // Edge should be deleted because its source node is gone
     });
 
+    test('should reject invalid node ids before Store.remove sees null (#11698)', async () => {
+        expect(() => db.removeNode(null)).toThrow(/non-empty string node id/);
+        expect(() => db.removeNode(undefined)).toThrow(/non-empty string node id/);
+        expect(() => db.removeNode('')).toThrow(/non-empty string node id/);
+    });
+
     test('should persist nodes and edges properly using SQLite storage adapter', async () => {
         // Clean out previous runs
         if (fs.existsSync(dbPath)) fs.unlinkSync(dbPath);
@@ -142,6 +148,18 @@ test.describe('Neo.ai.graph.Database', () => {
         expect(reloadDb.edges.get('e1').type).toBe('KNOWS');
 
         reloadDb.destroy();
+    });
+
+    test('SQLite storage rejects invalid node ids before persistence (#11698)', async () => {
+        let storage = Neo.create(SQLite, { dbPath });
+        await storage.initAsync();
+
+        try {
+            expect(() => storage.addNodes([{ id: null, label: 'Broken', properties: {} }])).toThrow(/non-empty string id/);
+            expect(() => storage.addNodes([{ label: 'Broken', properties: {} }])).toThrow(/non-empty string id/);
+        } finally {
+            storage.destroy();
+        }
     });
 
     test('SQLite storage enables foreign_keys pragma — source-side Edges cascade-delete (#10856)', async () => {

--- a/test/playwright/unit/ai/services/memory-core/GraphService.spec.mjs
+++ b/test/playwright/unit/ai/services/memory-core/GraphService.spec.mjs
@@ -132,6 +132,11 @@ test.describe('Neo.ai.services.memory-core.GraphService', () => {
         expect(task2.weight).toBe(0.8);
     });
 
+    test('removeNodes rejects invalid node ids before Database.removeNode null path (#11698)', async () => {
+        expect(() => GraphService.removeNodes([null])).toThrow(/invalid node id/);
+        expect(() => GraphService.removeNodes(['ValidNode', undefined])).toThrow(/invalid node id/);
+    });
+
     test('preBriefSession should hydrate episodic context through getNeighbors semanticVectorId', async () => {
         await GraphService.upsertNode({id: 'EpicA', name: 'Roadmap Planner'});
         await GraphService.upsertNode({id: 'MemoryA', name: 'Session Summary', semanticVectorId: 'summary-vector-1'});

